File: src/spacetraders/application/contracts/commands/deliver_contract.py
"""Deliver contract cargo command"""
from dataclasses import dataclass

from pymediatr import Request, RequestHandler

from ....domain.shared.contract import Contract, Delivery, ContractTerms
from ....ports.outbound.repositories import IContractRepository
from ....ports.outbound.api_client import ISpaceTradersAPI


@dataclass(frozen=True)
class DeliverContractCommand(Request[Contract]):
    """Command to deliver cargo for a contract"""
    contract_id: str
    ship_symbol: str
    trade_symbol: str
    units: int
    player_id: int


class DeliverContractHandler(RequestHandler[DeliverContractCommand, Contract]):
    """Handler for DeliverContractCommand"""

    def __init__(
        self,
        contract_repository: IContractRepository,
        api_client_factory
    ):
        self._contract_repo = contract_repository
        self._api_client_factory = api_client_factory

    async def handle(self, request: DeliverContractCommand) -> Contract:
        """
        Handle deliver contract cargo command

        Args:
            request: Command with delivery details

        Returns:
            Updated contract entity

        Raises:
            ValueError: If contract not found
        """
        # Get API client for player
        api_client = self._api_client_factory(request.player_id)

        # Call API to deliver cargo
        response = api_client.deliver_contract(
            request.contract_id,
            request.ship_symbol,
            request.trade_symbol,
            request.units
        )

        # Load contract from database
        contract = self._contract_repo.find_by_id(
            request.contract_id,
            request.player_id
        )

        if not contract:
            raise ValueError(f"Contract {request.contract_id} not found")

        # Update delivery progress from API response
        api_deliveries = response['data']['contract']['terms']['deliver']

        # Create updated deliveries list
        updated_deliveries = []
        for delivery in contract.terms.deliveries:
            # Find matching delivery in API response
            matching_api = next(
                (d for d in api_deliveries if d['tradeSymbol'] == delivery.trade_symbol),
                None
            )

            if matching_api:
                # Update with API data
                updated_delivery = Delivery(
                    trade_symbol=delivery.trade_symbol,
                    destination=delivery.destination,
                    units_required=delivery.units_required,
                    units_fulfilled=matching_api['unitsFulfilled']
                )
            else:
                # Keep existing
                updated_delivery = delivery

            updated_deliveries.append(updated_delivery)

        # Create new contract with updated deliveries (immutable)
        terms = ContractTerms(
            deadline=contract.terms.deadline,
            payment=contract.terms.payment,
            deliveries=updated_deliveries
        )

        updated_contract = Contract(
            contract_id=contract.contract_id,
            faction_symbol=contract.faction_symbol,
            type=contract.type,
            terms=terms,
            accepted=contract.accepted,
            fulfilled=contract.fulfilled,
            deadline_to_accept=contract.deadline_to_accept
        )

        # Save updated contract
        self._contract_repo.save(updated_contract, request.player_id)

        return updated_contract

File: gobot/internal/application/contract/deliver_contract.go
package contract

import (
	"context"
	"fmt"

	"github.com/andrescamacho/spacetraders-go/internal/application/common"
	"github.com/andrescamacho/spacetraders-go/internal/domain/contract"
	"github.com/andrescamacho/spacetraders-go/internal/domain/player"
	"github.com/andrescamacho/spacetraders-go/internal/infrastructure/ports"
)

// DeliverContractCommand - Command to deliver cargo for a contract
type DeliverContractCommand struct {
	ContractID  string
	ShipSymbol  string
	TradeSymbol string
	Units       int
	PlayerID    int
}

// DeliverContractResponse - Response from deliver contract command
type DeliverContractResponse struct {
	Contract       *contract.Contract
	UnitsDelivered int
}

// DeliverContractHandler - Handles deliver contract commands
type DeliverContractHandler struct {
	contractRepo contract.ContractRepository
	apiClient    ports.APIClient
	playerRepo   player.PlayerRepository
}

// NewDeliverContractHandler creates a new deliver contract handler
func NewDeliverContractHandler(
	contractRepo contract.ContractRepository,
	apiClient ports.APIClient,
	playerRepo player.PlayerRepository,
) *DeliverContractHandler {
	return &DeliverContractHandler{
		contractRepo: contractRepo,
		apiClient:    apiClient,
		playerRepo:   playerRepo,
	}
}

// Handle executes the deliver contract command
func (h *DeliverContractHandler) Handle(ctx context.Context, request common.Request) (common.Response, error) {
	cmd, ok := request.(*DeliverContractCommand)
	if !ok {
		return nil, fmt.Errorf("invalid request type")
	}

	// 1. Get player token
	player, err := h.playerRepo.FindByID(ctx, cmd.PlayerID)
	if err != nil {
		return nil, fmt.Errorf("player not found: %w", err)
	}

	// 2. Load contract from repository
	contract, err := h.contractRepo.FindByID(ctx, cmd.ContractID, cmd.PlayerID)
	if err != nil {
		return nil, fmt.Errorf("contract not found: %w", err)
	}

	// 3. Validate delivery using domain logic (BEFORE calling API)
	if err := contract.DeliverCargo(cmd.TradeSymbol, cmd.Units); err != nil {
		return nil, err
	}

	// 4. Call API to deliver cargo
	deliveryData, err := h.apiClient.DeliverContract(
		ctx,
		cmd.ContractID,
		cmd.ShipSymbol,
		cmd.TradeSymbol,
		cmd.Units,
		player.Token,
	)
	if err != nil {
		return nil, fmt.Errorf("failed to deliver cargo: %w", err)
	}

	// 5. Update contract with actual delivery data from API response
	// The domain entity's DeliverCargo already updated the units,
	// but we need to sync with the API's actual response
	terms := contract.Terms()
	for i := range terms.Deliveries {
		// Find matching delivery in API response
		for _, apiDelivery := range deliveryData.Terms.Deliveries {
			if terms.Deliveries[i].TradeSymbol == apiDelivery.TradeSymbol {
				// Update the units fulfilled from API
				terms.Deliveries[i].UnitsFulfilled = apiDelivery.UnitsFulfilled
			}
		}
	}

	// 6. Save updated contract to repository
	if err := h.contractRepo.Add(ctx, contract); err != nil {
		return nil, fmt.Errorf("failed to save contract: %w", err)
	}

	return &DeliverContractResponse{
		Contract:       contract,
		UnitsDelivered: cmd.Units,
	}, nil
}
